packages/lib/server/service/hashedLinks.service.ts

```diff
@@ -0,0 +1,219 @@
+    if (!linkId || typeof linkId !== "string") {
+      throw new Error("Invalid link ID");
+    }
+
+    const hashedLink = await this.hashedLinksRepository.findLinkWithValidationData(linkId);
+
+    if (!hashedLink) {
+      throw new Error(ErrorCode.PrivateLinkExpired);
+    }
+
+    // Handle time-based expiration with timezone awareness
+    if (hashedLink.expiresAt) {
+      const hostTimezone = this.extractHostTimezone(hashedLink.eventType);
+
+      if (hostTimezone) {
+        // Use dayjs for timezone-aware comparison
+        const now = dayjs().tz(hostTimezone);
+        const expiration = dayjs(hashedLink.expiresAt).tz(hostTimezone);
+
+        if (expiration.isBefore(now)) {
+          throw new Error(ErrorCode.PrivateLinkExpired);
+        }
+      } else {
+        // Fallback to UTC comparison if no timezone available
+        const now = dayjs();
+        const expiration = dayjs(hashedLink.expiresAt);
+
+        if (expiration.isBefore(now)) {
+          throw new Error(ErrorCode.PrivateLinkExpired);
+        }
+      }
+      return hashedLink;
+    }
+
+    // Handle usage-based expiration
+    if (hashedLink.maxUsageCount && hashedLink.maxUsageCount > 0) {
+      if (hashedLink.usageCount >= hashedLink.maxUsageCount) {
+        throw new Error(ErrorCode.PrivateLinkExpired);
+      }
+
+      // Increment usage count with optimistic locking
+      try {
+        await this.hashedLinksRepository.incrementUsage(hashedLink.id, hashedLink.maxUsageCount);
+      } catch (updateError) {
+        throw new Error(ErrorCode.PrivateLinkExpired);
+      }
+    }
+
+    return hashedLink;
+  }
+
+  /**
+   * Checks if a user has permission to access a link
+   * @param link - Link object with event type information
+   * @param userId - The user ID to check permissions for
+   * @returns true if user has permission, false otherwise
+   */
+  async checkUserPermissionForLink(
+    link: { eventType: { teamId?: number | null; userId?: number | null } },
+    userId: number
+  ): Promise<boolean> {
+    if (!link?.eventType || !userId || userId <= 0) {
+      return false;
+    }
+
+    // If it's a user event type, check if user owns it
+    if (link.eventType.userId && link.eventType.userId !== userId) {
+      return false;
+    }
+
+    // If it's not a team event, user has permission
+    if (!link.eventType.teamId) {
+      return true;
+    }
+
+    // Check team membership
+    const membership = await this.prismaClient.membership.findFirst({
+      where: {
+        teamId: link.eventType.teamId,
+        userId,
+        accepted: true,
+      },
+      select: {
+        id: true,
+      },
+    });
+
+    return !!membership;
+  }
+
+  /**
+   * Extracts the host timezone from event type data
+   * @param eventType - Event type data with potential timezone information
+   * @returns Host timezone string or null if not found
+   */
+  private extractHostTimezone(eventType: EventTypeForTimezone): string | null {
```

Comment:
We need to use the event type owner's timezone here. Otherwise the link can be not expired but it might be expired in the host's timezone